XenAdmin/Wizards/CrossPoolMigrateWizard/CrossPoolMigrateWizard.cs

```diff
@@ -403,23 +410,24 @@ protected override void UpdateWizardContent(XenTabPage page)
                     AddAfterPage(m_pageCopyMode, m_pageIntraPoolCopy);
                     if (Helpers.ConnectionRequiresRbac(xenConnection))
                     {
-                        m_pageTargetRbac.AddApiMethodsCheck(xenConnection,
-                            VMCopyAction.StaticRBACDependencies,
-                            Messages.RBAC_INTRA_POOL_COPY_VM_BLOCKED);
+                        var message = m_vmMappings.Any(IsTemplate)
+                            ? Messages.RBAC_INTRA_POOL_COPY_TEMPLATE_BLOCKED
+                            : Messages.RBAC_INTRA_POOL_COPY_VM_BLOCKED;
+
+                        var rbac = new RbacMethodList();
+                        rbac.AddRange(VMCopyAction.StaticRBACDependencies);
+                        rbac.AddRange(VMCloneAction.StaticRBACDependencies);
+                        rbac.AddRange(SrRefreshAction.StaticRBACDependencies);
+
+                        m_pageTargetRbac.SetPermissionChecks(xenConnection,
+                            new WizardRbacCheck(message, rbac) {Blocking = true});
                         AddAfterPage(m_pageCopyMode, m_pageTargetRbac);
                     }
                 }
                 else
                 {
                     RemovePagesFrom(1);
                     AddAfterPage(m_pageCopyMode, m_pageDestination, m_pageStorage, m_pageFinish);
-                    if (Helpers.ConnectionRequiresRbac(xenConnection))
-                    {
-                        m_pageTargetRbac.AddApiMethodsCheck(xenConnection,
-                            VMCloneAction.StaticRBACDependencies,
-                            Messages.RBAC_CROSS_POOL_CLONE_VM_BLOCKED);
-                        AddAfterPage(m_pageCopyMode, m_pageTargetRbac);
-                    }

```

Comment:
Looking at it again, I realise I may need to add one more correction:
I think that blocking should depend on privileges only on the target connection, not on the source connection (xenConnection). I'll check again with Toolstack to be sure, but if this is the case, it's correct that I removed this check from here, because in the case of cross pool copy it won't allow the user to select a connection they may have privileges on. Which means we should remove the source connection check from lines 357 and below.

Reply:
I've checked with Toolstack, we need RBAC checks on both source and target connection, which means you're right that we can block it earlier. I'll rework it.